### PR TITLE
systemd-repart: Add an option to generate a JSON report

### PR DESCRIPTION
**Pull request by hunger**
Add a "SetupScript" directive to the Partition definition file.

Run these scripts after all partitions have been created and pass
the actual device node for that partition script.

### CODE REVIEW

**Comment by hunger** [comment]
Right now documentation and similar things are missing: I wanted to get feedback on the general idea and implementation before I invest more work into this patch.

This is not fully in line with Lennarts idea of adding support to create filesystems and copy files onto the new partitions, but I think this is the most flexible approach possible:-)

**Comment by poettering** [comment]
hmm, but what does this do that `systemd-dissect --mount … && some-script` couldn#t do equally well?

**Comment by hunger** [comment]
systemd-dissect is not mentioned here: https://www.freedesktop.org/software/systemd/man/ . I never heard about it, so maybe this is the same thing:-)

At this point systemd-repart can not format a partition, so how does systemd-dissect mount anything?

**Comment by hunger** [comment]
Reading the the code of systemd-dissect: I don't think this does very much the same thing.

I want to use the SetupScript to run cryptsetup, format and then populate the individual partitions created by systemd-repart. Throw in some subvolumes and loop-back filesystems here and there:-). I do not see how you can support all the different things a user might want to do with custom keywords in you partition's definition files and code in systemd-repart! I would appreciate if there was something to do a simple format (and I would use that!), but there is no way to do everything.

I tried running my scripts after systemd-repart was done, but it was really annoying to figure out which partition to work on! Systemd-repart might had skipped some partitions entirely due to their priority/size requirements. Other partititons were already set up from earlier runs, which requires some more logic in the scripts to detect that case and skip. Having systemd-repart trigger the scripts for newly created partitions and providing the script with the correct device node is very convenient:-)

**Comment by hunger** [comment]
So what is the verdict? Is this a direction in which you want to go? I'll write some documentation then:-)

Do you have a better idea to get arbitrarily complex partition setups, then please tell. I am open to implement something else!

**Comment by poettering** [comment]
So, I definitely want to have the following high-level operations in repart:

1. Encrypt=yes (only LUKS)
2. Format=ext4|xfs|btrfs|fat
3. CopyBlock=/some/rawfile
4. CopyFiles=/some/directory/tree

And these should be applied *before* we write the partition table, so that the partitions pop into existance fully populated. Either the partitions dont exist or they exist and are fully set up, but never "half setup"m never exist but not populated. This makes me quite weary regarding your patch: it does stuff after the fact.

systemd-dissect currently has no man page, but we should add it and make it official. It's a tool that can dissect images following the discoverable partition spec. You can mount images with that.

My goal is to make repart good enough to do the basic setup in atomic fashion as described above, and then permit users to make further adjustments by doing "systemd-dissect --mount". repart should be smart enough to set everything up enough so that "systemd-dissect --mount" just works afterwards.

I am not a fan of "callout script" interfaces. It's vague and usually just means that our tool should probably do more natively.

What precisely do you need to do that doesn't fit under the four points above? (Note that the directory tree copy routine we already have in place is smart enough to duplicate btrfs subvols in the copy. hence, if your source tree uses btrfs subvolumes the copy will too)

By doing these things inside repart we have other benefits btw, beyond the atomic stuff. For example, we canmake things reasonably reproducible, because we can derive UUIDs for the file systems/volumes from our common seed and so on.

my primary usecase for items 1 + 2 from the list above are OS images that come with only immutable /usr, and that in the initrd will add in an encrypted root partition, that is later linked to the TPM. My primary usecases for 3 + 4 are to make repart a simple, replicating installer, that is smart enough to even install sd-boot on a system, by creating an esp if needed and the copying in all files really needed.

**Comment by hunger** [comment]
I am fully on board with all four commands you want to add. Those make a lot of sense to have since that is what 99% of the users will need.

I am firmly in the 1% here that would want to do more than this and am not overly surprised to get my use-case rejected.

My build server creates a lot of disk images consisting of this:
* A EFI partition with systemd-boot plus a signed kernel/initrd/os-release/... combo in EFI/Linux
* A squashfs filesystem with /usr (Is there anything better than squashfs out there, need to check that at some point)
* A dm-verity partition for the squashfs filesystem.

The interesting part is that I want to set up a bare metal machine from scratch with nothing but such an image.

For that I need to set up all the disks in the machine: I need to partition *all* HDDs and get them into a state where everything I need to boot up my image is in place.

Right now I am formatting partitions, setting up an EFI partition, setting up encryption, defining PVs/VGs/LVs with thin-provisioning and then run some cluster filesystems on top of some of those LVs. I also restore backups on some other LVs and checkout code into others. 

I seriously doubt that I can shoehorn all that into systemd-repart -- nor do I expect most of this be even useful for 99% of the users out there! I also doubt that systemd-dissect will help me with most of this.

I test all this in qemu -- where I have a couple of virtual drives created from scratch, have those partitioned and set up and copy my image file onto those drives in the right place and expect the system to boot up fully.

I tried to run systemd-repart and run a script afterwards to do the actual setup. That works, but it is pretty painful to figure out after the fact which partitions were actually created (and which ones already existed), and to distinguish between the different partitions (some may be skipped due to limited space, etc.), so the idea was born to let systemd-repart run a setup script for each partition it just created.

If you find it unacceptable to have systemd-repart trigger the scripts itself, would you mind if I make it write a machine-readable report of what it did? Which partitions are there, which file is responsible for which parttiion, which were created/grown/left alone, which device node goes with which partition, that kind of thing? The UI reports all that already, but it is not easy to extract this information from a script. Maybe "--report=foo.json" (or whatever format would fit) to trigger systemd-repart writing out a report file?

**Comment by topimiettinen** [comment]
On #14451 @bluca and I discussed about various approaches for constructing squashfs images and actually I mentioned repartd as a possible declarative solution.

So I'd also propose 1. to include verity support and 2. to include squashfs. With those additions this seems close to being useful for my case and perhaps also closer to @hunger's case. I'm also using LVM for the images.

It would be interesting to compare various scripts used to construct the images. Mine are at https://github.com/topimiettinen/make-images.

**Comment by poettering** [comment]
So there are three usecases for repart: 

1. to run on first boot and expand the image we live on, i.e. take possession of the disk
2. as an installer, where you boot from an installer medium and then a disk image is streamed onto the disk to install things to
3. in OS image generation on some build server

Now of these I think the first two items are the ones that matter most. i.e. the 3rd item will be called as one step of many in build scripts, i.e. are something developers script and developers maintain locally, as part of their build system, in a single instance. It doesn't have to be as streamlined I think. Now, mksquashfs and verity are probably stuff that is primarily relevant in case 3, i.e. where you generate an image on the build system, then finally glue it together in a GPT parttable using systemd-repart. I don't think there's much benefit of doing the mksquashfs and verity building inside of systemd-repart for that: it's totally fine to build those things first, and then pass the generated images to systemd-repart via CopyBlocks= to glue them together. There's very little gained if systemd-repart would do that for you. I see no use of ever running mksquashfs or verity building on the system things are deployed to, these tools are as far as I can see clearly developers tools run on build systems only.

So I am pretty sure mksquashfs/verity make stuff should live outside of systemd-repart, and I think it probably should just be done ahead of time and then be glued together with repart.

LVM stuff should really stay outside of things too. Not a fan anyway, and I think it would just make things too complex for what I have in mind.

boot loader installation is probably something "bootctl install"/"bootctl update" should do, and it probably should learn --image= so that you can point it directly to a disk image and it will do the right thing. Installing/updating boot loaders probably needs a bit more knowledge than just dropping in a few files: it might need version checks, EFI variable changes and so on. 

What's also interesting to realize, I figure for cases 1 + 2 systemd-repart will work on physical block devices, but for cases 3 more likely on regular files it fallocate()s. Interestingly you (@hunger) appear to use the tool mostly for case 3 but your patch only works for physical block devices afaics, right? (because the partition names your patch generates are useless if we operate on files that do not exist as kernel partition devices). So if we'd go for the script approach like your current PR we'd have to find a solution for that, i.e. probably pass in an fd plus offset or so instead of a partition name. Or synthesize a loopback device first if needed, or so.

So, I like your JSON idea a lot more. We already output what we are doing in a pretty table. Our table formatting code already optionally outputs things as JSON  (look for `table_print_json()`). I think if we'd add a new switch "--json" or so that changes the table output to use the JSON mode you'd already be at 80% of what you need, right? The remaining 20% would then be to tweak the table a bit so that you actually have all the data you need in it. Would that be a workable approach for you?

**Comment by poettering** [comment]
@topimiettinen so, there's mkosi which is the tool i myself use for building discoverable images. It can do a number of things, but isn't where I want it yet. Right now it uses "sfdisk" to set up the partition table, it's my intention to eventually use systemd-repart for that.

mkosi can already do secureboot and verity stuff. It should be relatively straight-forward to do more of that, for example sign the root hash and so on.

**Comment by hunger** [comment]
@poettering : I have neither of the 3 use cases you list nor do I need help with discoverable partitions, dm-verity, squashfs or secure boot. All that works just fine already:-)

I have use-case 4: "Partition drives and do stuff with those partitions". I deal with either physical drives or with virtual drives for qemu or vmware. Currently I am mostly using qemu-nbd for its support of the different disk formats via a set of block devices -- which makes it very easy to support VM disks and "real" hardware drives.

I have an image created specifically for each of my systems. Those images know a logical disk layout they expect to find on the machine. Now I want to deploy that image onto a physical or virtual machine.

* I boot the server from a USB stick/network drive.

* I tell it which system image to use for the machine and which of the physical disks it sees correspond to which logical disk the custom system image expects.

* Magic happens and the physical disks are fully set up.

* The system image is deployed to the right place on the physical disks.

* A reboot will take me into the system image I originally created.

All that works for me -- except for the "Magic happens" part:-) Systemd-repart comes into play there: The image file already contains systemd-repart definition files for all partitions on all logical drives it knowns. The declarative syntax is really nice for that! My problem is mostly with continuing after the partitioning -- especially when optional partitions are in the mix somewhere :-)

I do not at all mind scripting LVM setup and whatever else I need, I just need a reliable way to know which of the partitions were created and which device node those ended up with.

This patch does that by having systemd-repart trigger the scripts itself and by having systemd-repart pass the device node to each script it starts.

I would be equally happy to have a machine readable report in some format that I can parse from a wide range of languages so that I can do whatever is necessary myself.

I'll try to find some time to change this patch to just write some report file then.

**Comment by hunger** [comment]
Sorry, I meant to push close this PR and open a new one, but github went ahead and just updated this one:-/

Anyway: The latest patch adds a --json=(pretty|short|off) switch and generates an activity report. I would appreciate feedback on the data in the report. Anything missing?

I am not particularly happy with this though: The json output is mixed into the "normal" output of systemd-repart, which makes it hard to parse. I would try for a different. I would suggest to add a "--report-file=<file>" that will write the json data into the file. Is this in line with the other systemd commands?

If a --report-file switch is added, do we still need the --json one to select the JSON parameters?

**Comment by mrc0mmand** [comment]
The CentOS CI fail is #16105 - unrelated.

File: src/partition/repart.c [hunk 1]
```diff
@@ -1699,6 +1702,59 @@ static int context_dump_partitions(Context *context, const char *node) {
         return 0;
 }
 
+static int generate_report(Context *context, const char *node) {
```
Comment (poettering) [inline src/partition/repart.c]: hmm, so it was my hope that we could just reuse the table that context_dump_partitions() outputs, and maybe add a few more columns that we incude in the JSON table but maybe turn off in the text version (i.e. using text_set_display())

File: src/partition/repart.c [hunk 2]
```diff
@@ -1699,6 +1702,59 @@ static int context_dump_partitions(Context *context, const char *node) {
...
+                return log_oom();
+
+        LIST_FOREACH(partitions, p, context->partitions) {
+                _cleanup_free_ char *partname = NULL;
+                char uuid_buffer[ID128_UUID_STRING_MAX];
+                const char *label = NULL;
+                const char *activity = NULL;;
+
+                if (p->dropped)
+                        activity = "dropped";
```
Comment (poettering) [inline src/partition/repart.c]: the dropped ones should probably never be shown

File: src/partition/repart.c (hunk 1 above)
Comment (hunger) [reply src/partition/repart.c]: There are two reasons why I did not do that:
* The current table has graphical elements that do provide value in the output and that I would want to keep. So I would need to create a different table for JSON anyway (e.g. with separate columns for old_value and new_value instead of merging both values into one string with an arrow between.
* The existing table summarizes the changes systemd-repart *wants* to do, while I want to have a report of changes that systemd-repart actually *did*.

File: src/partition/repart.c (hunk 2 above)
Comment (hunger) [reply src/partition/repart.c]: Why? It was listed *somewhere*, so it should IMHO show up in the report:-)

File: src/partition/repart.c (hunk 1 above)
Comment (hunger) [reply src/partition/repart.c]: The table is way more flexible then I first understood and the return value is a good indicator of whether or not systemd-repart succeeded, so my arguments are void.

I changed this now.

File: src/partition/repart.c (hunk 2 above)
Comment (hunger) [reply src/partition/repart.c]: I removed the dropped partitions now.

**Comment by hunger** [comment]
I am not sure why the bionic-i386 test fails: From all I understand it does not even build systemd-repart!

**Comment by hunger** [comment]
@keszybz: Could you maybe take a look at my patch? I think it is in a pretty good state now and even includes documentation -- even though a test that as far as I can tell does not even build systemd-repart fails.

File: man/systemd-repart.xml [hunk 3]
```diff
@@ -265,6 +265,15 @@
...
       </varlistentry>
 
+      <varlistentry>
+        <term><option>--json=</option><replaceable>MODE</replaceable></term>
+
+	<listitem><para>Shows output formatted as JSON. Expects one of <literal>short</literal> (for the
+	shortest possible output without any redundant whitespace or line breaks), <literal>pretty</literal>
+	(for a pretty version of the same, with indentation and line breaks) or <literal>off</literal> (to turn
+	off json output).</para></listitem>
+      </varlistentry>
```
Comment (poettering) [inline man/systemd-repart.xml:275]: please follow usual indentation

File: src/partition/repart.c [hunk 4]
```diff
@@ -1619,12 +1622,24 @@ static int context_dump_partitions(Context *context, const char *node) {
         Partition *p;
         int r;
 
-        t = table_new("type", "label", "uuid", "file", "node", "offset", "raw size", "size", "raw padding", "padding");
+        if (!arg_json && context->n_partitions == 0) {
+                puts("Empty partition table.");
```
Comment (poettering) [inline src/partition/repart.c]: hmm, log_info() probably, i.e. if you redirect output to some file this shouldn't show up i am sure...

File: src/partition/repart.c [hunk 5]
```diff
@@ -1619,12 +1622,24 @@ static int context_dump_partitions(Context *context, const char *node) {
...
+                return 0;
+        }
+
+        t = table_new("type", "label", "uuid", "file", "node", "offset", "old size", "raw size", "size", "old padding", "raw padding", "padding", "activity");
+
```
Comment (poettering) [inline src/partition/repart.c]: please keep function call and the immediate error handling together, i.e. remove the empty line here

File: src/partition/repart.c [hunk 6]
```diff
@@ -1634,8 +1649,14 @@ static int context_dump_partitions(Context *context, const char *node) {
                 char uuid_buffer[ID128_UUID_STRING_MAX];
                 const char *label;
 
+                const char *activity = NULL;;
```
Comment (poettering) [inline src/partition/repart.c]: please merge with previous line, and there's a double `;;` here

File: src/partition/repart.c [hunk 7]
```diff
@@ -1634,8 +1649,14 @@ static int context_dump_partitions(Context *context, const char *node) {
...
+
                 if (p->dropped)
                         continue;
+                else if (p->current_size == UINT64_MAX)
```
Comment (poettering) [inline src/partition/repart.c]: `else if` → `if`. The if checks are entirely unrelated after all, and given the `continue` the first branch is final in this block anyway...

File: man/systemd-repart.xml (hunk 3 above)
Comment (hunger) [reply man/systemd-repart.xml:275]: Oh, sorry for this, I did not notice the tab my editor added!

File: src/partition/repart.c [hunk 8]
```diff
@@ -2799,6 +2825,8 @@ static int help(void) {
                "     --definitions=DIR    Find partitions in specified directory\n"
                "     --seed=UUID          128bit seed UUID to derive all UUIDs from\n"
                "     --size=BYTES         Grow loopback file to specified size\n"
+               "     --json=(pretty|short|off)\n"
```
Comment (keszybz) [inline src/partition/repart.c]: Please remove the parenthesis. We just say `--option=a|b|c`.

**Comment by keszybz** [comment]
I'll set the green label: @poettering already like this, and my changes on top are trivial.

**Comment by hunger** [comment]
Thanks @poettering and @keszybz for the review (and fixes:-)

**Comment by mrc0mmand** [comment]
The CentOS CI fail is unrelated (#15978)...

**Comment by hunger** [comment]
I fail to see how the CentOS Ci (arch in KVM) failure is related to my change. Do I need to fix something there?

**Comment by hunger** [comment]
Ping. The next release is out, can this go in now?
